var/spack/repos/builtin/packages/dpcpp/package.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+import os
+
+from spack import *
+
+
+class Dpcpp(CMakePackage):
+    """Data Parallel C++ compiler: Intel's implementation of SYCL programming model"""
+
+    homepage = 'https://intel.github.io/llvm-docs/'
+    git = 'https://github.com/intel/llvm.git'
+
+    version('develop', branch='sycl')
+    version('stable', commit='7981e6027a8cb43028be55d19f12cb1e30a28f5a')
+
+    maintainers = ['ravil-mobile']
+    variant('cuda', default=False, description='switch from OpenCL to CUDA')
+    variant('rocm', default=False, description='switch from OpenCL to ROCm')
+    variant('rocm-platform', default='AMD', values=('AMD', 'NVIDIA'), multi=False, description='choose ROCm backend')
+    variant('arm', default=False, description='build ARM support rather than x86')
+    variant('esimd-cpu', default=False, description='build with ESIMD_CPU support')
+    variant('assertions', default=False, description='build with assertions')
+    variant('docs', default=False, description='build Doxygen documentation')
+    variant('werror', default=False, description='Treat warnings as errors')
+    variant('shared', default=False, description='Build shared libraries')
+    variant('lld', default=False, description='Use LLD linker for build')
+
+    depends_on('cmake@3.16.2:', type='build')
+    depends_on('ninja@1.10.0:', type='build')
+
+    depends_on('cuda@10.2.0:10.2.999', when='+cuda')
+
+    # NOTE: AMD HIP needs to be tested; it will be done in the next update
+    # depends_on('cuda@10.2.0:10.2.999', when='rocm-platform=NVIDIA', type='build')
+    # depends_on('hip@4.0.0:', when='+rocm', type='build')
+
+    build_targets = ['deploy-sycl-toolchain']
+    install_targets = ['deploy-sycl-toolchain']
+
+    root_cmakelists_dir = 'llvm'
+
+    patch('cuda-backend.patch', when='@stable +cuda')
```

Comment:
Hi, @bader and @alalazo,

I just want to mention that I am not a developer of dpcpp. I am just a user who always wanted to have a reproducible installation of this compiler. 

To be honest, I cannot figure out versioning of dpcpp (see, [here](https://github.com/intel/llvm/releases) and [there](https://github.com/intel/llvm/tags)). There are lot's of tags (nightly builds) and it is not clear what is stable and what is not. `stable` version in Spack refers to just a nightly build (that I took a month ago) that didn't fail any test. I just needed a reference to something which is not changing e.g., not a tip of the master branch.  However, I've just noticed that there is a **really** stable release ([this one](https://github.com/intel/llvm/releases/tag/2021-07)). So I will change the reference. 

Talking about patch, honestly the solution came from [this thread](https://github.com/intel/llvm/issues/4307). I didn't go into details what caused the problem. But because I believe that the bug will be fixed in the future I didn't apply the patch to the develop branch.

Reply:
Hello, the patch doesn't work anymore with https://github.com/intel/llvm/commit/3c7971d77de6a4f7cdf5831f4fb909391afd89f3

Reply:
but the bug is still present

Reply:
@ravil-mobile can you try `nvptx64-nvidia-cuda--sm_50` ?

Reply:
Hi, @Michoumichmich,

Do you mean I can take `dpcpp` from the tip of the master, compile it, compile my application (or a test application) with `-fsycl -fsycl-targets=nvptx64-nvidia-cuda--sm_50`, and it will work without the patch?

What about [this release](https://github.com/intel/llvm/releases/tag/2021-07)? Do I still need a patch for this? I mean a Spack package can have a `develop` version which points to the tip of the master but there should be at least one stable version (or at least something which people can reference as a stable version).

Reply:
All the issues I had with dpcpp that needed that patch are gone on the master branch and the target `nvptx64-nvidia-cuda--sm_whatever` and `nvptx64-nvidia-cuda`. I don't know if the release version works without the patch, but it works with it. Maybe you could take the release and patch on top of it the commit mentioned ? It could work. I can't try right now. 
Here's a code that was breaking before: https://github.com/Michoumichmich/SYCL-Hashing-Algorithms

Reply:
My apologies for being confusing

Reply:
@ravil-mobile Hi, I just checked and the release you mentioned still needs the patch, else it fails with CUDA_ERROR_NOT_FOUND

Reply:
> Hi, I just checked and the release you mentioned still needs the patch, without it the executable fails with CUDA_ERROR_NOT_FOUN

Ok, thanks. Btw, I tried the tip of the master (3 days ago) without the patch, and it resulted in lot's of warning during computation. Is it supposed to be like that?

Reply:
@ravil-mobile do you mean compilation ? What are they ?

Reply:
BTW, if you need setup instructions for the whole oneAPI/DNN/MKL, I've written one here https://github.com/Michoumichmich/oneAPI-setup-script

Reply:
Hi, @Michoumichmich. I tried `2021-09` release and it seems to that everything is working now.